Classes/EventStore/AbstractEventSourcedRepository.php

```diff
@@ -59,23 +57,19 @@ public function __construct()
 
     /**
      * @param string $identifier
-     * @return AggregateRootInterface
+     * @return EventSourcedAggregateRootInterface
      * @throws AggregateRootNotFoundException
      */
-    public function findByIdentifier($identifier)
+    final public function findByIdentifier($identifier)
     {
-        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
-        try {
-            $eventStream = $this->eventStore->get($streamName);
-        } catch (EventStreamNotFoundException $e) {
-            return null;
-        }
-
         if (!class_exists($this->aggregateClassName)) {
             throw new AggregateRootNotFoundException(sprintf("Could not reconstitute the aggregate root %s because its class '%s' does not exist.", $identifier, $this->aggregateClassName), 1474454928115);
         }
 
-        $aggregateRoot = unserialize('O:' . strlen($this->aggregateClassName) . ':"' . $this->aggregateClassName . '":1:{s:13:"' . chr(0) . '*' . chr(0) . 'identifier";s:36:"' . $identifier . '";}');
+        $streamName = $this->streamNameResolver->getStreamNameForAggregateTypeAndIdentifier($this->aggregateClassName, $identifier);
+        $eventStream = $this->eventStore->get($streamName);
+
+        $aggregateRoot = unserialize('O:' . strlen($this->aggregateClassName) . ':"' . $this->aggregateClassName . '":2:{s:13:"' . chr(0) . '*' . chr(0) . 'identifier";s:' . strlen($identifier) . ':"' . $identifier . '";s:10:"' . chr(0) . '*' . chr(0) . 'version";i:' . (string)$eventStream->getVersion() . ';}');
```

Comment:
How about we set identifier/version in the "fromHistory()"/"reconstituteFromEventStream()" method of the aggregate instead of adding to this hacky instanciation that attempts to create a semi-valid aggregate with identifier and a version - which does not really hold because no events have affected it yet.

What I imagine is doing this:

```
if ($identifier !== $eventStream[0]->getAggregateIdentifier()) {
   throw new WrongAggregateEventStreamException(...);
}
$aggregateRoot = ... // just create an empty instance without new by some means
$aggregateRoot->fromHistory($eventStream);
```

Then in `fromHistory`:
```
public function fromHistory($eventStream)
{
   if ($this->identifier) {
      throw new AggregateAlreadyReconstitutedException(...);
   }
   $this->identifier = $eventStream[0]->getAggregateIdentifier();
   foreach ($eventStream as $eventWithMetadata) {
     if ($eventWithMetadata->getAggregateIdentifier() !== $this->identifier) {
        throw new WrongAggregateEventException(...);
     }
     if ($eventWithMetadata->getVersion() !== $this->version + 1) {
        throw new OutOfOrderEventException(...);
     }
     $this->apply($eventWithMetadata->getEvent());
     $this->version = $eventWithMetadata->getVersion();
   }
}
```

That way:
- the EventStream becomse fully decoupled from the Aggregate, stream version could be something different from the aggregate version even (what is the [aggregate] version of an event stream created by a cross-aggregate filter or from a ProcessManager?) IMO the stream version only has a meaning inside the stream itself.

- Guaranteeing aggregate version constraints becomes a concern of the aggregate itself again - it can use any versioning semantics it wants and store and retrieve them from the event metadata only without any dependency to the infrastructure

- `fromHistory` (or whatever we name it) effectively becomes a constructor method. We could theoretically even make it static and let the instanciation become its concern, then an Aggregate implementation could simply use a private constructor instead of the hack we currently do - and I would suggest that to be our default implementation (we already discussed using private constructor and advising static named constructor methods)